src/main/kotlin/com/github/shirleh/datacollection/DataCollectionHandler.kt

```diff
@@ -154,4 +155,42 @@ object DataCollectionHandler {
 
         logger.exit()
     }
+
+    /**
+     * Collects voice data from the incoming [events].
+     */
+    suspend fun collectVoiceData(events: Flow<VoiceStateUpdateEvent>) = events.collect { saveVoiceData(it) }
+
+    private fun saveVoiceData(event: VoiceStateUpdateEvent) {
+        logger.entry(event)
+
+        val voiceState = event.current
+
+        val guildId = voiceState.guildId.asString()
+        val guildMemberId = voiceState.userId.asString()
+
+        val channelId = voiceState.channelId
+            .map { it.asString() }
+            .orElse(
+                event.old
+                    .flatMap { old -> old.channelId.map { it.asString() } }
+                    .orElse("")
+            )
```

Comment:
That looks :ok_hand: ! I'll change one little thing: `orElse(...)` to `orElseGet { ... }` (reasoning in [recipe 5](https://blogs.oracle.com/javamagazine/12-recipes-for-using-the-optional-class-as-its-meant-to-be-used)).

Since it's a tiny nitpick, I'll commit that change from this side and approve the PR. :smile: 

```suggestion
            .orElseGet {
                event.old
                    .flatMap { old -> old.channelId.map { it.asString() } }
                    .orElse("")
            }
```